src/display/node_stream.js

```diff
@@ -0,0 +1,367 @@
+    method: 'GET',
+    headers,
+  };
+}
+
+class PDFNodeStreamFullReader extends BaseFullReader {
+  constructor(stream) {
+    super(stream);
+
+    let handleResponse = (response) => {
+      this._headersCapability.resolve();
+      this._setReadableStream(response);
+
+      let { allowRangeRequests, suggestedLength, } =
+      validateRangeRequestCapabilities({
+        getResponseHeader: (name) => {
+          return this._readableStream.headers[name.toLowerCase()];
+        },
+        isHttp: stream.isHttp,
+        rangeChunkSize: this._rangeChunkSize,
+        disableRange: this._disableRange,
+      });
+
+      if (allowRangeRequests) {
+        this._isRangeSupported = true;
+      }
+      this._length = suggestedLength;
+    };
+
+    this._request = null;
+    if (this._url.protocol === 'http:') {
+      this._request = http.request(createRequestOptions(
+        this._url, stream.httpHeaders), handleResponse);
+    } else {
+      this._request = https.request(createRequestOptions(
+        this._url, stream.httpHeaders), handleResponse);
+    }
+
+    this._request.on('error', (reason) => {
+      this._errored = true;
+      this._reason = reason;
+      this._headersCapability.reject(reason);
+    });
+    this._request.end();
+  }
+}
+
+class PDFNodeStreamRangeReader extends BaseRangeReader {
+  constructor(stream, start, end) {
+    super(stream);
+
+    this._httpHeaders = {};
+    for (let property in stream.httpHeaders) {
+      let value = stream.httpHeaders[property];
+      if (typeof value === 'undefined') {
+        continue;
+      }
+      this._httpHeaders[property] = value;
+    }
+    this._httpHeaders['Range'] = `bytes=${start}-${end - 1}`;
+
+    this._request = null;
+    if (this._url.protocol === 'http:') {
+      this._request = http.request(createRequestOptions(
+        this._url, this._httpHeaders), (response) => {
+          this._setReadableStream(response);
+        });
+    } else {
+      this._request = https.request(createRequestOptions(
+        this._url, this._httpHeaders), (response) => {
+          this._setReadableStream(response);
+        });
+    }
+
+    this._request.on('error', (reason) => {
+      this._errored = true;
+      this._reason = reason;
+    });
+    this._request.end();
+  }
+}
+
+class PDFNodeStreamFsFullReader extends BaseFullReader {
+  constructor(stream) {
+    super(stream);
+
+    this._setReadableStream(fs.createReadStream(this._url.path));
+
+    fs.lstat(this._url.path, (error, stat) => {
+      if (error) {
+        this._errored = true;
+        this._reason = error;
+        this._headersCapability.reject(error);
+        return;
+      }
+      this._length = stat.size;
+      this._headersCapability.resolve();
+    });
+  }
+}
+
+class PDFNodeStreamFsRangeReader extends BaseRangeReader {
+  constructor(stream, start, end) {
+    super(stream);
+
+    this._setReadableStream(
+      fs.createReadStream(this._url.path, { start, end, }));
```

Comment:
I still have `(node:90529) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): Error: Bad end offset: 65537`. Node.js operates with `end` being inclusive. Make it `end: end - 1,`